citation-matching/citation-matching-core-code/src/main/java/pl/edu/icm/coansys/citations/BestMatchedCitationPicker.java

```diff
@@ -0,0 +1,116 @@
+        
+        JavaPairRDD<TextWithBytesWritable, Text> bestMatchedCitations = citDocPairs
+                .mapPartitionsToPair(citAndDocParitition -> {
+                    
+                    SimilarityMeasurer similarityMeasurer = new SimilarityMeasurer(SimilarityMeasurer.advancedFvBuilder());
+                    
+                    return calculateSimilarityAndFilter(similarityMeasurer, citAndDocParitition);
+                })
+                .reduceByKey((docWithSim1, docWithSim2) -> (docWithSim1.getSimilarity() > docWithSim2.getSimilarity()) ? docWithSim1 : docWithSim2)
+                .mapValues(docWithSim -> new Text(docWithSim.getSimilarity() + ":" + docWithSim.getId()));
+
+        return bestMatchedCitations;
+    }
+    
+    
+    //------------------------ PRIVATE --------------------------
+    
+    private List<Tuple2<TextWithBytesWritable, IdWithSimilarity>> calculateSimilarityAndFilter(SimilarityMeasurer similarityMeasurer, Iterator<Tuple2<TextWithBytesWritable, TextWithBytesWritable>> citAndDocIterable) {
+        
+        List<Tuple2<TextWithBytesWritable, IdWithSimilarity>> citDocIdPairsWithSimilarity = Lists.newArrayList();
+        
+        
+        while(citAndDocIterable.hasNext()) {
+            
+            Tuple2<TextWithBytesWritable, TextWithBytesWritable> citAndDoc = citAndDocIterable.next();
+            
+            double similarity = calculateSimilarity(similarityMeasurer, citAndDoc._1, citAndDoc._2);
+            
+            if (similarity >= MIN_SIMILARITY) {
+                citDocIdPairsWithSimilarity.add(createCitDocWithSimilarityPair(citAndDoc._1, citAndDoc._2, similarity));
+            }
+        }
+        
+        return citDocIdPairsWithSimilarity;
+    }
+    
+    private Tuple2<TextWithBytesWritable, IdWithSimilarity> createCitDocWithSimilarityPair(
+            TextWithBytesWritable citationWritable, TextWithBytesWritable documentWritable, double similarity) {
+        
+        return new Tuple2<TextWithBytesWritable, IdWithSimilarity>(
+                citationWritable, 
+                new IdWithSimilarity(similarity, documentWritable.text().toString()));
+    }
+    
+    private double calculateSimilarity(SimilarityMeasurer similarityMeasurer,
+            TextWithBytesWritable citationWritable, TextWithBytesWritable documentWritable) {
+        
+        MatchableEntity citation = MatchableEntity.fromBytes(citationWritable.bytes().copyBytes());
+        MatchableEntity document = MatchableEntity.fromBytes(documentWritable.bytes().copyBytes());
+        
+        double similarity = similarityMeasurer.similarity(citation, document);
+        
+        return similarity;
+    }
+
+    
+    private static class IdWithSimilarity {
+        
+        private double similarity;
+        private String id;
+        
+        public IdWithSimilarity(double similarity, String id) {
```

Comment:
It would be better to have id, similarity constructor (Id--- with --- similarity)